toolchain/lower/file_context.cpp

```diff
@@ -136,17 +137,21 @@ auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id)
     return nullptr;
   }
 
-  // Don't lower unused functions.
-  if (function.return_slot == SemIR::Function::ReturnSlot::NotComputed) {
-    return nullptr;
-  }
+  // TODO: Consider tracking whether the function has been used, and only
+  // lowering it if it's needed.
+
+  // TODO: Pass in a specific ID for generic functions.
+  const auto specific_id = SemIR::GenericInstanceId::Invalid;
 
-  const bool has_return_slot = function.has_return_slot();
+  const auto return_info = function.GetReturnInfo(sem_ir(), specific_id);
+  CARBON_CHECK(return_info.is_valid()) << "Should not lower invalid functions.";
```

Comment:
Isn't this duplicating the check in `has_return_slot`?

Reply:
Yes, but it produces a more useful check failure message.